openedx/core/lib/gating/api.py

```diff
@@ -299,3 +305,143 @@ def get_gated_content(course, user):
                 {'id': user.id}
             )
         ]
+
+
+def is_gate_fulfilled(course_key, gating_content_key, user_id):
+    """
+    Determines if a prerequiste section specified by gating_content_key
+    has any unfulfilled milestones.
+
+    Arguments:
+        course_key (CourseUsageLocator): Course locator
+        gating_content_key (BlockUsageLocator): The locator for the section content
+        user_id: The id of the user
+
+    Returns:
+        Returns True if section has no unfufilled milestones or is not a prerequiste.
+        Returns False otherwise
+    """
+    gating_milestone = get_gating_milestone(course_key, gating_content_key, "fulfills")
+    if not gating_milestone:
+        return True
+
+    unfulfilled_milestones = [
+        m['content_id'] for m in find_gating_milestones(
+            course_key,
+            None,
+            'requires',
+            {'id': user_id}
+        ) if m['namespace'] == gating_milestone['namespace']
+    ]
+    return not unfulfilled_milestones
+
+
+def compute_is_prereq_met(content_id, user_id, recalc_on_unmet=False):
+    """
+    Returns true if the prequiste has been met for a given milestone.
+    Will recalculate the subsection grade if specified and prereq unmet
+
+    Arguments:
+        content_id (BlockUsageLocator): BlockUsageLocator for the content
+        user_id: The id of the user
+        recalc_on_unmet: Recalculate the grade if prereq has not yet been met
+
+    Returns:
+        tuple: True|False,
+        prereq_meta_info = { 'url': prereq_url|None, 'display_name': prereq_name|None}
+    """
+    course_id = content_id.course_key
+
+    # if unfullfilled milestones exist it means prereq has not been met
+    unfulfilled_milestones = milestones_helpers.get_course_content_milestones(
+        course_id,
+        content_id,
+        'requires',
+        user_id
+    )
+
+    prereq_met = not unfulfilled_milestones
+    prereq_meta_info = {'url': None, 'display_name': None}
+
+    if prereq_met or not recalc_on_unmet:
+        return prereq_met, prereq_meta_info
+
+    milestone = unfulfilled_milestones[0]
+    student = User.objects.get(id=user_id)
+    store = modulestore()
+
+    with store.bulk_operations(course_id):
+        subsection_usage_key = UsageKey.from_string(_get_gating_block_id(milestone))
+        subsection = store.get_item(subsection_usage_key)
+        prereq_meta_info = {
+            'url': reverse('jump_to', kwargs={'course_id': course_id, 'location': subsection_usage_key}),
+            'display_name': subsection.display_name
+        }
+
+        if not subsection.visible_to_staff_only:
```

Comment:
Why is it necessary to do this access control check here?  Typically, we'd like to keep all access-checking within the BlockTransformer code (and within [has_access](https://github.com/edx/edx-platform/blob/e7c869f87a5d5eafec227665a5bd40bad7eae5d8/lms/djangoapps/courseware/access.py#L97) for older code).

The call to `get_course_blocks` below should already do the `visible_to_staff_only` check for non-staff users.

Reply:
@nasthagiri 
I added this check because was getting an error during this (unlikely) scenario:
- if you setup a section to be gated on another section whose visibility is "Hide Entire Subsection" which is not included in grade calculations.
- when you click on the gated section it tries to recalc the grade of the hidden/staff only section and blows up with this stack trace: https://pastebin.ubuntu.com/26348794/
- get_course_block seems to return a valid subsection_structure object but it blows up in creation of SubsectionGradeFactory()

Is there an attribute of subsection_structure that I can check and skip SubsectionGradeFactory() in that case?

Reply:
@nasthagiri 
I updated the code to use any(subsection_structure) before instantiating SubsectionGradeFactory() and that seems to work